virtual-programs/shapes.folk

```diff
@@ -1,111 +1,301 @@
+      angle $angle color $color filled $filled thickness $thickness layer $layer
+      
   } else {
-      Wish to draw a shape with sides 2 \
-          center $p radius $radius radians $angle \
-          color $color filled $filled layer $layer
+    set radius [dict_getdef $options radius 50]
+    
+    Wish to draw a shape with sides 4 center $center radius $radius \
+      angle $angle color $color filled $filled thickness $thickness layer $layer
   }
 }
+
+# Pass through options for "an" version
 When /someone/ wishes /p/ draws an /shape/ with /...options/ {
   Wish $p draws a $shape with {*}$options
 }
 
+When /someone/ wishes /p/ draws a rect with width /w/ height /h/ {
+  Wish $p draws a rect with width $w height $h
+}
+
+When /someone/ wishes /p/ draws a /shape/ with radius /rad/ {
+  Wish $p draws a $shape with radius $rad
+}
+
+When /someone/ wishes /page/ draws a set of points /points/ with /...options/ & /page/ has region /r/ {
+  set radius [dict_getdef $options radius 5]
+  set color [dict_getdef $options color white]
+  set filled [dict_getdef $options filled true]
+  set thickness [dict_getdef $options thickness 2]
+  set layer [dict_getdef $options layer 0]
+  
+  lassign [region centroid $r] cx cy
+  set angle [region angle $r]
+  set center [list $cx $cy]
+  
+  if {[dict exists $options offset]} {
+    set offset [dict get $options offset]
+    set offset [process_offset $offset $r]
+    set center [vec2 add $center [vec2 rotate $offset $angle]]
+  }
+  
+  foreach point $points {
+    set pointPos [vec2 add $center [vec2 rotate $point $angle]]
+    
+    Wish to draw a circle with center $pointPos radius $radius thickness $thickness \
+      color $color filled $filled layer $layer
+  }
+}
+
+When /someone/ wishes /page/ draws a polyline /points/ with /...options/ & /page/ has region /r/ {
+  set color [dict_getdef $options color white]
+  set thickness [dict_getdef $options thickness 2]
+  set layer [dict_getdef $options layer 0]
+  set dashed [dict_getdef $options dashed false]
+  set dashlength [dict_getdef $options dashlength 20]
+  set dashoffset [dict_getdef $options dashoffset 0]
+  
+  lassign [region centroid $r] cx cy
+  set angle [region angle $r]
+  set center [list $cx $cy]
+  
+  if {[dict exists $options offset]} {
+    set offset [dict get $options offset]
+    set offset [process_offset $offset $r]
+    set center [vec2 add $center [vec2 rotate $offset $angle]]
+  }
+  
+  set transformedPoints {}
+  foreach point $points {
+    lappend transformedPoints [vec2 add $center [vec2 rotate $point $angle]]
+  }
+  
+  if {$dashed} {
+    Wish to draw a dashed stroke with points $transformedPoints color $color width $thickness \
+      dashlength $dashlength dashoffset $dashoffset layer $layer
+  } else {
+    Wish to draw a stroke with points $transformedPoints color $color width $thickness layer $layer
+  }
+}
+
 Claim $this has demo {
```

Comment:
how are these demos meant to be used in practice? 

this works well and is kind of fun but not sure what your  intent was -- is there a specific tool somewhere?

```
When virtual-programs/shapes.folk has demo /d/ {
  eval $d
}
```

Reply:
We have `has demo` in `sprites.folk` and `editor.folk` (although `editor` is a weird one to have ...), I'm imagining a booklet (or dial or something?) that allows you to view all the demos, I like the idea that the system is self-documenting in this way, need to actually prototype what using these will look like though.